org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ProjectUtils.java

```diff
@@ -235,6 +235,72 @@ public static boolean isOnSourcePath(IPath sourcePath, IJavaProject project) thr
 		return false;
 	}
 
+	public static boolean setInvisibleProjectOutputPath(IJavaProject javaProject, String outputPath, IProgressMonitor monitor) throws CoreException {
+		IProject project = javaProject.getProject();
+		if (isVisibleProject(project)) {
+			return false;
+		}
+
+		if (outputPath == null) {
+			outputPath = "";
+		} else {
+			outputPath = outputPath.trim();
+		}
+		
+		if (new org.eclipse.core.runtime.Path(outputPath).isAbsolute()) {
+			throw new CoreException(new Status(IStatus.ERROR, IConstants.PLUGIN_ID, "The output path must be a relative path to the workspace."));
+		}
+
+		if (StringUtils.isEmpty(outputPath)) {
+			// blank means using default output path
+			javaProject.setOutputLocation(project.getFolder("bin").getFullPath(), monitor);
+			return true;
+		}
+
+		outputPath = WORKSPACE_LINK + IPath.SEPARATOR + outputPath;
+		IPath outputFullPath = project.getFolder(outputPath).getFullPath();
+		IClasspathEntry[] existingEntries = javaProject.getRawClasspath();
+		List<IClasspathEntry> newEntries = new ArrayList<>();
+		for (IClasspathEntry entry : existingEntries) {
+			if (entry.getEntryKind() != IClasspathEntry.CPE_SOURCE) {
+				newEntries.add(entry);
+			} else {
+				IPath entryPath = entry.getPath();
+				if (entryPath.equals(outputFullPath)) {
+					throw new CoreException(new Status(IStatus.ERROR, IConstants.PLUGIN_ID,
+							"The output path cannot equal to source folder path, please provide a new path instead."));
+				} else if (entryPath.isPrefixOf(outputFullPath)) {
+					List<IPath> exclusionPatterns = new ArrayList<>(Arrays.asList(entry.getExclusionPatterns()));
+					IPath excludePath = outputFullPath.makeRelativeTo(entryPath);
+					boolean hasExcluded = false;
+					for (IPath exclusionPattern : exclusionPatterns) {
+						if (exclusionPattern.isPrefixOf(excludePath)) {
+							hasExcluded = true;
+							break;
+						}
+					}
+					if (!hasExcluded) {
+						exclusionPatterns.add(excludePath.addTrailingSeparator());
+					}
+					newEntries.add(JavaCore.newSourceEntry(entryPath, exclusionPatterns.toArray(new IPath[0])));
+					continue;
+				} else if (outputFullPath.isPrefixOf(entryPath)) {
+					throw new CoreException(new Status(IStatus.ERROR, IConstants.PLUGIN_ID,
+							"The specified output path contains source folders, please provide a new path instead."));
+				} else {
+					newEntries.add(entry);
+				}
+			}
+		}
+
+		try {
+			javaProject.setRawClasspath(newEntries.toArray(new IClasspathEntry[0]), outputFullPath, monitor);
```

Comment:
When i'm trying this feature, i found this operation may delete my existing files in the output folder.

Reproduce steps:
- Open a standalone folder as below.
```
 |── app
 │   ...
 │   └── build
 |          └── run.script
 |   └── src
 |          └── Hello.java
```
- set `build` in the setting as the output path. -> The language server will auto delete my existing files in this folder. This is risky without the user's awareness.

IMO, we should ask user for confirmation if the specified output path is not empty. Or just let the operation fail directly, and provide a specific error message to explain why.

Reply:
Thanks for catching that, I changed to throw a `CoreException` instead.

Reply:
Good catch on this case. It might be good to check setting a non-writeable output folder as most of the checks seem to be around existence only, but seems like even more of a corner case. Do we plan this for 0.71.0 ?

Reply:
@rgrunber there are some new concerns about the UX on our side. The current solution has two problems:
- it will automatically apply to JDT.LS just after the user finishes typing
- it allow the user to save into the global scoped setting, which might pollute other unmanaged folders

A rough thinking right now is to change to expose a command, which is to set a folder to the compiled output path, and expose the command at the client extension. This is an initial idea, if you have any feedback, please let me know :)